lxd/certificates.go

```diff
@@ -711,30 +707,92 @@ func certificatePatch(d *Daemon, r *http.Request) response.Response {
 
 	clientType := clusterRequest.UserAgentClientType(r.Header.Get("User-Agent"))
 
-	return doCertificateUpdate(d, *oldEntry, fingerprint, req.Writable(), clientType, r)
+	return doCertificateUpdate(d, *oldEntry, req.Writable(), clientType, r)
 }
 
-func doCertificateUpdate(d *Daemon, dbInfo db.Certificate, fingerprint string, req api.CertificatePut, clientType clusterRequest.ClientType, r *http.Request) response.Response {
+func doCertificateUpdate(d *Daemon, dbInfo db.Certificate, req api.CertificatePut, clientType clusterRequest.ClientType, r *http.Request) response.Response {
 	if clientType == clusterRequest.ClientTypeNormal {
 		reqDBType, err := db.CertificateAPITypeToDBType(req.Type)
 		if err != nil {
 			return response.BadRequest(err)
 		}
 
 		// Convert to the database type.
-		cert := db.Certificate{
-			// Read-only fields.
+		dbCert := db.Certificate{
 			Certificate: dbInfo.Certificate,
 			Fingerprint: dbInfo.Fingerprint,
+			Restricted:  req.Restricted,
+			Projects:    req.Projects,
+			Name:        req.Name,
+			Type:        reqDBType,
+		}
+
+		if req.Certificate != "" && dbInfo.Certificate != req.Certificate {
+			// Only allow certificate updates by its owner and the admin
+			if !rbac.UserIsAdmin(r) && r.TLS != nil {
```

Comment:
Actually this check is wrong, if the user isn't an admin and isn't using a TLS cert, this would incorrectly allow the change...

I think we need the if to just be `if !rbac.UserIsAdmin(r)` and then inside that block, check if `r.TLS == nil` and if it is, immediately fail.